immunopepper/filter.py

```diff
@@ -79,189 +75,167 @@ def _get_exon_junction_dict(metadata_list, strand):
-        j-th vertex are connected.
-    vertex_id_tuple: triple-elements tuple. (v1,v2,v3) represents the vertex id for the junction pair.
-
-    Returns
-    -------
-    int or list of int depends on it being a 2-vertex junction of 3-vertex junction.
-
+    :param junction_flag: 2-d array with shape (number_of_vertices, number_of_vertices). junction_flag[i,j] == 1 iff
+       vertices i and j are connected
+    :param vertex_ids: tuple of vertex ids defining junction pairs
+    :return:
+       np.nan if vertex_id_tuple contains a np.nan
+       1 if any junction in param:`vertex_id_tuple` is connected in param:`junction_flag`, 0 otherwise
     """
-    if np.nan in vertex_id_tuple:
+    if np.nan in vertex_ids:
         return np.nan
-    for i in range(len(vertex_id_tuple) - 1):
-        if junction_flag[vertex_id_tuple[i], vertex_id_tuple[i+1]]:
+    for i in range(len(vertex_ids) - 1):
+        if junction_flag[vertex_ids[i], vertex_ids[i + 1]]:
             return 1
     return 0
 
 
-def junction_is_in_given_list(splicegraph, vertex_id_list, strand, junction_list):
-    """Check if the intron is in the user provided list of junctions"""
+def is_intron_in_junction_list(splicegraph: Splicegraph, vertex_ids: list[int], strand: str,
+                               junction_list: list[str]):
+    """
+    Check if the intron defined by vertex_ids is in the user provided list of junctions
+    :param splicegraph: Spladder splice graph
+    :param vertex_ids: list vertex coordinates in splicegraph defining the intron (typically 2 coordinates)
+    :param strand: '+' for direct, '-' for reverse strand
+    :param junction_list: list of junctions to check the intron against
+    :return: np.nan if junction_list is None or if vertex_ids contains a np.nan
+             1 if the intron is in junction_list, 0 if not
+    """
 
-    if np.nan in vertex_id_list or junction_list is None:
+    if np.nan in vertex_ids or junction_list is None:
         return np.nan
 
     vertex_coord_list = splicegraph.vertices
     if strand == '-':
-        vertex_id_list = vertex_id_list[::-1] # in negative strand case, the vertex id list is [5,3] instead of [3,5]
-    for i in range(len(vertex_id_list) - 1):
-        junc_comb = ':'.join([str(vertex_coord_list[1, vertex_id_list[i]]),  str(vertex_coord_list[0, vertex_id_list[i+1]]), strand])
-        if junc_comb in junction_list:
+        vertex_ids = vertex_ids[::-1]  # in negative strand case, the vertex id list is [5,3] instead of [3,5]
+    for i in range(len(vertex_ids) - 1):
+        junction_comb = ':'.join(
+            [str(vertex_coord_list[1, vertex_ids[i]]), str(vertex_coord_list[0, vertex_ids[i + 1]]), strand])
+        if junction_comb in junction_list:
             return 1
     return 0
 
 
+# TODO(dd): this function smells. It takes nicely structured data and massages it into an undecipherable jumble. Why?
```

Comment:
I concur with this question. Not sure what the added value is.

Reply:
Yes I think that the goal here is to output 
1 line for each
 kmer (resp) peptide __  gene{1};gene{i};expression{n}__  expressionsegment{1};expressionsegment{i};expressionsegment{n}__  isCrossJunctionFlag{1}isCrossJunctionFlag{i}isCrossJunctionFlag{n}__ expressionjunction{1};expressionjunction{i};expressionjunction{n}__  
 
 Where __ represents a tab in the output file 
 And {1,..i,..n} are unique instances of a given property for this kmer/peptides ever observed
 (currently it is rather "ever observed within a gene")

Reply:
This function is ugly, yes. What should have been done in the first place is to completely change the OutputKmer (resp OutputBackground, OutputMetadata ) data structures.

Reply:
I am mostly worried about using dicts, sets and strings to represent information, which makes the code hard to read and maintain. Having a class that describes and stores the output result would be better, but I need more context to fix this. Maybe later.